src/main/java/carpetextra/utils/BlockPlacer.java

```diff
@@ -2,117 +2,167 @@
 
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.AbstractRailBlock;
+import net.minecraft.block.ChestBlock;
 import net.minecraft.block.ComparatorBlock;
-import net.minecraft.block.DispenserBlock;
+import net.minecraft.block.DetectorRailBlock;
 import net.minecraft.block.FacingBlock;
-import net.minecraft.block.GlazedTerracottaBlock;
 import net.minecraft.block.HorizontalFacingBlock;
-import net.minecraft.block.ObserverBlock;
-import net.minecraft.block.PistonBlock;
+import net.minecraft.block.PoweredRailBlock;
+import net.minecraft.block.RailBlock;
 import net.minecraft.block.RepeaterBlock;
 import net.minecraft.block.StairsBlock;
 import net.minecraft.block.TrapdoorBlock;
+import net.minecraft.block.WallMountedBlock;
 import net.minecraft.block.enums.BlockHalf;
+import net.minecraft.block.enums.ChestType;
 import net.minecraft.block.enums.ComparatorMode;
+import net.minecraft.block.enums.RailShape;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemPlacementContext;
+import net.minecraft.state.property.DirectionProperty;
+import net.minecraft.state.property.Property;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Direction;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 
-public class BlockPlacer
-{
+public class BlockPlacer {
+    private static Boolean HasDirectionProperty(BlockState state) {
+        //malilib code
+        for (Property<?> prop : state.getProperties()) {
+            if (prop instanceof DirectionProperty) {
+                return true;
+            }
+        }
+        return false;
+    }
+
```

Comment:
I'd remove this and instead save the result of `getFirstDirectionProperty` in a field (since it will be needed later anyway) and check `!= null`